### PR TITLE
Eigen calculation

### PR DESCRIPTION
**Pull request by NikhilShahGH**
The logic for calculation of the Support Degree Matrix and its Eigen Values and Vectors has been implemented.

### CODE REVIEW

File: src/SensorFusionAlgorithm.c [hunk 1]
```diff
@@ -3,22 +3,55 @@
...
+			            }
+			            temp = exp(-(temp));
+			            d[i][j] =temp;
+			            e[k]= temp;
+			            printf("%lf\t",e[k]);
+			            k++;
+			            if(k%size==0)
+			            	printf("\n");
+			        }
+			    }
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c]: Idendation

File: src/SensorFusionAlgorithm.c [hunk 2]
```diff
@@ -3,22 +3,55 @@
...
-        }
-    }
-    return d;
+	double sensorinputs[] = {1, 12, 6, 4, 7, 6, 8};
+	 int i,j,x,k=0;
+			    int size = sizeof(sensorinputs)/sizeof(sensorinputs[0]);
+			    printf("size %d\n", size);
+			    double d[size][size];
+			    double *e = malloc(sizeof(double)*(size*size));
+			    for(i=0; i<size;i++)
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c]: Memory not deallocated for (*e)

File: src/SensorFusionAlgorithm.c [hunk 3]
```diff
@@ -3,22 +3,55 @@
...
+			        }
+			    }
+			    gsl_matrix_view m = gsl_matrix_view_array (e, size, size);
+			    gsl_vector *eval = gsl_vector_alloc (size);
+			    gsl_matrix *evec = gsl_matrix_alloc (size, size);
+			    gsl_eigen_symmv_workspace * w = gsl_eigen_symmv_alloc (size);
+			    gsl_eigen_symmv (&m.matrix, eval, evec, w);
+			    gsl_eigen_symmv_free (w);
+			    gsl_eigen_symmv_sort (eval, evec, GSL_EIGEN_SORT_ABS_ASC);
+			    {
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c]: According to paper and project statement eigen values need to be in descending order

File: src/SensorFusionAlgorithm.c [hunk 4]
```diff
@@ -1,24 +1,127 @@
...
+ *  array. This will be later changed to dynamically handle inputs from
+ *  a function that reads csv file and outputs an array containing sensor
+ *  inputs.
+ */
+double* sdm_calculator(double sensorinputs[], int size){
+
+    int i,j,k=0;
+	printf("size %d\n", size);
+	double d[size][size];
+	double *dmatrix = malloc(sizeof(double)*(size*size));
```
Comment (karanbirsandhu) [inline src/SensorFusionAlgorithm.c:26]: double * dmatrix variable memory has not been deallocated.

**Review by akaljotdhillon — COMMENTED** [review]
1. Indendation
2. Memory allocated need to be deallocated after use
3. Size variable should be descriptive

File: src/SensorFusionAlgorithm.c [hunk 5]
```diff
@@ -1,24 +1,127 @@
...
+double* eigen_value_calculation(double* e, int size){
+	int i;
+	gsl_matrix_view m = gsl_matrix_view_array (e, size, size);
+	gsl_vector *eval = gsl_vector_alloc (size);
+	gsl_matrix *evec = gsl_matrix_alloc (size, size);
+	gsl_eigen_symmv_workspace * w = gsl_eigen_symmv_alloc (size);
+	gsl_eigen_symmv (&m.matrix, eval, evec, w);
+	gsl_eigen_symmv_free (w);
+	gsl_eigen_symmv_sort (eval, evec, GSL_EIGEN_SORT_ABS_DESC);
+	double *eval_i = malloc(sizeof(double)*(size));
```
Comment (karanbirsandhu) [inline src/SensorFusionAlgorithm.c:59]: eval_i memory has not been deallocated.

File: src/SensorFusionAlgorithm.c [hunk 6]
```diff
@@ -1,24 +1,127 @@
...
+		gsl_eigen_symmv (&m.matrix, eval, evec, w);
+		gsl_eigen_symmv_free (w);
+		gsl_eigen_symmv_sort (eval, evec, GSL_EIGEN_SORT_ABS_DESC);
+		double *evec_i = malloc(sizeof(double)*(size));
+		for(i=0;i<size;i++)
+		{
+			evec_i[i]= gsl_matrix_get(evec, i, column);
+			//printf("%g\n", evec_i[i]);
+		}
+		return evec_i;
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c:82]: memory not deallocated for *evec_i

File: src/SensorFusionAlgorithm.c [hunk 7]
```diff
@@ -1,24 +1,127 @@
...
+	for(int i=0;i <size;i++) {
+		list_of_alphas[i] = eval_i[i]/sum_of_evals;
+		printf("list of alpha = %lf\n",list_of_alphas[i]*100);
+	}
+	return list_of_alphas;
+}
+
+
+double* compute_phi(double list_of_alphas[], int size){
+	double *list_of_phi = malloc(sizeof(double)*(size));
```
Comment (karanbirsandhu) [inline src/SensorFusionAlgorithm.c:118]: list_of_phi memory has not been deallocated.

File: src/SensorFusionAlgorithm.c [hunk 8]
```diff
@@ -1,24 +1,127 @@
...
+ *                            accumulated contribution rates i.e., phi
+ */
+double* compute_alpha(double eval_i[],int size){
+	double *list_of_alphas = malloc(sizeof(double)*(size));
+	double sum_of_evals = 0.0;
+	for(int i=0; i<size; i++){
+		sum_of_evals += eval_i[i];
+	}
+	for(int i=0;i <size;i++) {
+		list_of_alphas[i] = eval_i[i]/sum_of_evals;
```
Comment (karanbirsandhu) [inline src/SensorFusionAlgorithm.c:110]: list_of_alphas memory has not been deallocated.

File: src/SensorFusionAlgorithm.c [hunk 9]
```diff
@@ -1,24 +1,127 @@
...
+		printf("list of alpha = %lf\n",list_of_alphas[i]*100);
+	}
+	return list_of_alphas;
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c:113]: list_of_alphas variable name should be more descriptive

File: src/SensorFusionAlgorithm.c [hunk 10]
```diff
@@ -1,24 +1,127 @@
...
+
+
+double* compute_phi(double list_of_alphas[], int size){
+	double *list_of_phi = malloc(sizeof(double)*(size));
+	list_of_phi[0] = list_of_alphas[0];
+		for(int i=1; i<size;i++){
+			list_of_phi[i] = list_of_phi[i-1] + list_of_alphas[i];
+			printf("list of phi = %lf\n",list_of_phi[i]*100);
+		}
+		return list_of_phi;
```
Comment (akaljotdhillon) [inline src/SensorFusionAlgorithm.c:124]: list_of_phis variable name should be more descriptive

**Review by karanbirsandhu — COMMENTED** [review]
Please free the allocated memory to avoid memory conflicts.

**Review by akaljotdhillon — COMMENTED** [review]
Please add ripudaman92 for review as well

**Comment by rishabhjiresal** [comment]
This branch is outdated. The changes requested are being implemented on an updated branch - FusionAlgorithm. We will create a pull request in the newer branch once the requested changes have been resolved.

**Pull request closed without merging**